File: code/python/ch14/practice/challenge_01_2d_prefix_sum.py
"""
Challenge 1: 2D Prefix Sum and Range Query
============================================
Chapter 14: Prefix Sums — The Power of Pre-computation

PROBLEM
-------
Given a 2D matrix and multiple queries, each specifying a rectangle
[r1, c1, r2, c2], return the sum of all elements in that rectangle
for each query. Use 2D prefix sums for efficient querying.

INPUT FORMAT
------------
First line: two integers rows and cols.
Next rows lines: cols space-separated integers.
Next line: integer q (number of queries).
Next q lines: four integers r1, c1, r2, c2.

OUTPUT FORMAT
-------------
A list of rectangle sums.

CONSTRAINTS
-----------
- 1 <= rows, cols <= 500
- -10^6 <= matrix[i][j] <= 10^6
- 1 <= q <= 10^5
- 0 <= r1 <= r2 < rows, 0 <= c1 <= c2 < cols

EXAMPLES
--------
Input:
  3 3
  1 2 3
  4 5 6
  7 8 9
  2
  0 0 2 2
  1 1 2 2
Output: [45, 28]

Input:
  1 1
  5
  1
  0 0 0 0
Output: [5]

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(matrix: list[list[int]], queries: list[list[int]]) -> list[int]:
    """Return list of rectangle sums for each query."""
    pass  # TODO: Replace this with your solution



# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    rows, cols = map(int, input().split())
    matrix = []
    for _ in range(rows):
        matrix.append(list(map(int, input().split())))
    q = int(input())
    queries = []
    for _ in range(q):
        queries.append(list(map(int, input().split())))
    print(solve(matrix, queries))

File: code/cpp/ch14/practice/challenge_01_2d_prefix_sum.cpp
/*
 * Challenge 1: 2D Prefix Sum and Range Query
 * Chapter 14: Prefix Sums — The Running Total Trick
 *
 * PROBLEM: Build 2D prefix sum, answer rectangle queries [r1,c1,r2,c2].
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */

#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

vector<long long> solve(vector<vector<int>> matrix, vector<vector<int>> queries) {
    // TODO: Replace this with your solution
    return vector<long long>(queries.size(), 0);
}

// -- Do not change anything below this line --------------------------
int main() {
    int rows, cols;
    cin >> rows >> cols;
    vector<vector<int>> matrix(rows, vector<int>(cols));
    for (int i = 0; i < rows; i++)
        for (int j = 0; j < cols; j++)
            cin >> matrix[i][j];
    int q;
    cin >> q;
    vector<vector<int>> queries(q, vector<int>(4));
    for (int i = 0; i < q; i++)
        cin >> queries[i][0] >> queries[i][1] >> queries[i][2] >> queries[i][3];
    auto result = solve(matrix, queries);
    for (int i = 0; i < (int)result.size(); i++)
        cout << result[i] << (i < (int)result.size()-1 ? " " : "\n");
    return 0;
}
